### DISCUSSION TITLE
Question: how to work with the WebSocketClient from the 'com.soywiz.korio.net.ws' package?

### DISCUSSION THREAD

**Discussion by benifest** [korlibs/korge#1234] [Q&A] [answered]
Hello guys. I can not find any example how to work with the WebSocketClient.
It looks very simple and easy to use.

Here is my implementation:
```
WebSocketClient("ws://localhost:8080/ws") {
            onOpen.add { println("onOpen") }
            onError.add { println("onError") }
            onStringMessage.add { println("onStringMessage") }
            onClose.add { println("onClose") }
        }
```
But it does not work. Or works not correctly. First of all when a connection opens i don't receive the onOpen event. And also when a connection closes i don't receive the onClose event. I am 100% sure that my server works correctly. Because my JavaScript implementation of the WebSocket works fine.

I am using the 3.1.0 version of KorGE.

**Comment by soywiz** [comment]
Can you try with 3.4.0 that is the latest version, just in case it was fixed already?

https://github.com/korlibs/korge/blob/326a0ff93ce7bd06f0e88708643426aed38fc680/korio/src/commonTest/kotlin/com/soywiz/korio/net/ws/RawSocketWebSocketClientTest.kt#L72

Are you trying in the JS target, right?

**Reply by benifest** [reply]
No. I am trying in the JVM target. 
When i said my JavaScript implementation works fine i meant i have an other project written on js\ts.

I am talking about this: https://github.com/korlibs/korge/blob/326a0ff93ce7bd06f0e88708643426aed38fc680/korio/src/commonMain/kotlin/com/soywiz/korio/net/ws/WebSocketClient.kt

**Reply by soywiz** [reply]
For me this is working as expected:

![Screenshot 2023-01-30 at 11 14 12](https://user-images.githubusercontent.com/570848/215449212-97d46799-55b8-445d-91cd-14f805f996a1.png)

**Reply by benifest** [reply]
Oh, okay. This line does the work: `while (true) com.soywiz.korio.async.delay(1000.seconds)`

**Comment by soywiz** [comment, marked as answer]
The thing is that once connected, the program continues, and it finishes when reaching the end even if there are pending connections. If you don't have an event loop, you should wait until a condition triggers.

You could do for example:
```kotlin
suspend fun main() {
    val closed = Signal<Unit>()

    val ws = WebSocketClient("ws://localhost:8000/ws") {
        onOpen.add { println("onOpen") }
        onError.add { println("onError") }
        onStringMessage.add { println("onStringMessage") }
        onClose.add { println("onClose"); closed() }
    }

    closed.waitOne()
    // while (true) delay(1.seconds) // alternatively we can wait forever
}
```
